gcc/rust/typecheck/rust-hir-type-check-expr.cc

```diff
@@ -1055,9 +1055,11 @@ TypeCheckExpr::visit (HIR::MethodCallExpr &expr)
 			     expr.get_method_name ().get_segment ());
   if (candidates.empty ())
     {
+      rich_location richloc (line_table, expr.get_method_name ().get_locus ());
+      richloc.add_fixit_replace ("method not found");
       rust_error_at (
-	expr.get_method_name ().get_locus (),
-	"failed to resolve method for %<%s%>",
+	richloc, ErrorCode::E0599,
+	"no method %qs name found in the current scope",
 	expr.get_method_name ().get_segment ().as_string ().c_str ());
       return;
     }
```

Comment:
I tried by comparing expression to get the `Rust::HIR::Expr::ExprType`, but every time it emits the same `ExprType` i.e., `enum Rust::HIR::Expr::ExprType::MethodCall = 9`, on all these testcases:

- [`gcc/testsuite/rust/compile/cfg2.rs`](https://github.com/Rust-GCC/gccrs/blob/master/gcc/testsuite/rust/compile/cfg2.rs)
- [`gcc/testsuite/rust/compile/issue-2139.rs`](https://github.com/Rust-GCC/gccrs/blob/master/gcc/testsuite/rust/compile/issue-2139.rs)
- [`gcc/testsuite/rust/compile/issue-2190-1.rs`](https://github.com/Rust-GCC/gccrs/blob/master/gcc/testsuite/rust/compile/issue-2190-1.rs)
- [`gcc/testsuite/rust/compile/method1.rs`](https://github.com/Rust-GCC/gccrs/blob/master/gcc/testsuite/rust/compile/method1.rs)

Reply:
Yes, I am referrring to the `ExprType` of the object on which the method is called:

https://github.com/Rust-GCC/gccrs/blob/3196ca35261f475c92b655dd943c2fc3f9c6d42e/gcc/rust/hir/tree/rust-hir.h#L290

I am trying to emit messages like this,

if we are using `struct` for method call, then, we should emit like,
```C++
no method %qs name found for struct in the current scope
```
and same for other `ExprTypes`

Reply:
ah, what you want is to get the `ExprType` of the method call's *receiver*. So when you have `a.foo()`, `a` is the receiver, and `foo` is the method. If you look at the `ExprType` of the receiver, you can get messages like these:

```rust
arthur@platypus ~/G/r/gccrs ((925e69ce))> build/gcc/crab1 2571.rs
2571.rs:2:23: error: no method ‘foo’ name found in the current scope for block [E0599]
    2 |     { let a = 15; a }.foo();
      |                       ^~~
      |                       method not found

Analyzing compilation unit
```

The code for this looks like this:

```c++
      auto receiver_kind = expr.get_receiver ()->get_expression_type ();
      switch (receiver_kind)
	{
	case HIR::Expr::Block:
	  name = "block";
	  break;
	/* ... */
```

Reply:
Yep, I was asking for this. Thanks :)

Reply:
Thanks, @CohenArthur for this suggestion. I tried using this approach. But it emit `enum Rust::HIR::Expr::ExprType::Path = 24` this enum for these examples:

- [`gcc/testsuite/rust/compile/cfg2.rs`](https://github.com/Rust-GCC/gccrs/blob/master/gcc/testsuite/rust/compile/cfg2.rs)
- [`gcc/testsuite/rust/compile/issue-2139.rs`](https://github.com/Rust-GCC/gccrs/blob/master/gcc/testsuite/rust/compile/issue-2139.rs)
- [`gcc/testsuite/rust/compile/issue-2190-1.rs`](https://github.com/Rust-GCC/gccrs/blob/master/gcc/testsuite/rust/compile/issue-2190-1.rs)
- [`gcc/testsuite/rust/compile/method1.rs`](https://github.com/Rust-GCC/gccrs/blob/master/gcc/testsuite/rust/compile/method1.rs)


There are many `HIR::Expr::`, which doesn't require this special edited message. So, I think the new message is best for all cases. WDYT?